codegen/src/main/java/software/amazon/awssdk/codegen/poet/waiters/JmesPathInterpreter.java

```diff
@@ -0,0 +1,306 @@
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.poet.waiters;
+
+import com.fasterxml.jackson.jr.stree.JrsBoolean;
+import com.fasterxml.jackson.jr.stree.JrsValue;
+import com.squareup.javapoet.CodeBlock;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.List;
+import software.amazon.awssdk.codegen.jmespath.component.AndExpression;
+import software.amazon.awssdk.codegen.jmespath.component.BracketSpecifier;
+import software.amazon.awssdk.codegen.jmespath.component.BracketSpecifierWithContents;
+import software.amazon.awssdk.codegen.jmespath.component.BracketSpecifierWithQuestionMark;
+import software.amazon.awssdk.codegen.jmespath.component.BracketSpecifierWithoutContents;
+import software.amazon.awssdk.codegen.jmespath.component.ComparatorExpression;
+import software.amazon.awssdk.codegen.jmespath.component.CurrentNode;
+import software.amazon.awssdk.codegen.jmespath.component.Expression;
+import software.amazon.awssdk.codegen.jmespath.component.ExpressionType;
+import software.amazon.awssdk.codegen.jmespath.component.FunctionArg;
+import software.amazon.awssdk.codegen.jmespath.component.FunctionExpression;
+import software.amazon.awssdk.codegen.jmespath.component.IndexExpression;
+import software.amazon.awssdk.codegen.jmespath.component.Literal;
+import software.amazon.awssdk.codegen.jmespath.component.MultiSelectHash;
+import software.amazon.awssdk.codegen.jmespath.component.MultiSelectList;
+import software.amazon.awssdk.codegen.jmespath.component.NotExpression;
+import software.amazon.awssdk.codegen.jmespath.component.OrExpression;
+import software.amazon.awssdk.codegen.jmespath.component.ParenExpression;
+import software.amazon.awssdk.codegen.jmespath.component.PipeExpression;
+import software.amazon.awssdk.codegen.jmespath.component.SliceExpression;
+import software.amazon.awssdk.codegen.jmespath.component.SubExpression;
+import software.amazon.awssdk.codegen.jmespath.component.SubExpressionRight;
+import software.amazon.awssdk.codegen.jmespath.component.WildcardExpression;
+import software.amazon.awssdk.codegen.jmespath.parser.JmesPathParser;
+import software.amazon.awssdk.codegen.jmespath.parser.JmesPathVisitor;
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.core.waiters.WaitersRuntime;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A code interpreter for JMESPath expressions.
+ *
+ * This can convert a JMESPath expression into a statement that executes against an {@link SdkPojo}.
```

Comment:
Can we add more detail in this javadoc? I was confused at first thinking the statements are jut basic Java expressions, but looks like the output is an intermediate representation that is further interpreted by the Waiters Runtime.

Reply:
Will do! As some history, I tried to just emit java expressions, but it was way too difficult. The waiters runtime made it much easier.